Production updates

ros3bud opened the pull request.

ros3bud commented:
@cgrayson, LGTY?

cgrayson commented:
Hope I'm understanding this right :smiley: I think you created this barely-different PR just as a place to put initial comments. I don't think can do line-comments here anyway, but here's my feedback.

Versium 

* link to Travis build in `README.md` is broken
* `package.json`
	* missing `repository.url`
	* `dependencies` includes unnecessary libs (all)
	* `devDependencies` doesn’t need `types` but does need `lodash`
* `common_lookup.coffee`
	* I’d call this just `common`, since it isn’t a lookup endpoint like the others
	* I think all the `validate` conditions can be simplified to `unless vars.lead.whatever?.valid` 
	* help me understand what the `if lead.email or null` on line 28 is meant to do?
	* in `create_event()`, perhaps all those `if parsedBody?.datafinder?`s could be factored out
	* the logic in `parseAddress()` seems like it could be simplified. Will there ever be `addr2` with no `addr1`? 
* `common_lookup_spec.coffee`
	* `emailIntegration` or `phoneIntegration` are `require`d but not used
	* in `describe ‘Validate’`, could add a passes-everything case
* `email_lookup.coffee`
	* seems like the response var `email_is_usable` should be boolean
	* what’s going on with the `alternate_emails` and `alternate_phones` arrays? those are populated with the given type, plus a timestamp in parens? 
	* response vars includes `time_stamp`, but the attribute in `common` is `data_timestamp` (same in `phone_lookup`)
* `email_lookup_spec.coffee`
	* it looks like some of these `body` values are repeated multiple times, making the code longer and a little less clear 
	* what is happening on line 169?
	* these tests cover successful data returned, but consider adding at least one for when looked-for data isn’t returned 

One last thing, about the naming: I wonder if it will be confusing that this company’s “email lookup” is meant to _find_ an email (if I’m understanding this correctly), whereas other “email lookup” integrations are meant to find data _by_ an email?

ros3bud commented:
link to Travis build in README.md is broken
>> updated

package.json
>> updated

- common_lookup.coffee
- I’d call this just common, since it isn’t a lookup endpoint like the others
>> updated
- I think all the validate conditions can be simplified to unless vars.lead.whatever?.valid
 >> I'm concerned that if I simplify this and there is a null value the field will be invalidated and I don't want the lookup to fail if that is the case.
- help me understand what the if lead.email or null on line 28 is meant to do?
>> After looking at that again, I don't think it's necessary to URI encode like that.
- in create_event(), perhaps all those if parsedBody?.datafinder?s could be factored out
>> I see that now, yes - makes sense. Updated.
- The logic in parseAddress() seems like it could be simplified. Will there ever be addr2 with no addr1?
>> I see your point and no, I don't think there would be. Updated.

common_lookup_spec.coffee
- emailIntegration or phoneIntegration are required but not used
>> Updated.
in describe ‘Validate’, could add a passes-everything case
>> Not sure what you mean by this.

email_lookup.coffee
- seems like the response var email_is_usable should be boolean
>> I used yes and no because that is the way the data comes back from Versium. Why do you think boolean is better? 
- what’s going on with the alternate_emails and alternate_phones arrays? those are populated with the given type, plus a timestamp in parens?
>> yes, that is correct. Those are emails or phone numbers which are also associated with the lead data. 
- response vars includes time_stamp, but the attribute in common is data_timestamp (same in phone_lookup)
>> thanks for seeing that. Updated.

email_lookup_spec.coffee
- it looks like some of these body values are repeated multiple times, making the code longer and a little less clear
>> I see what you mean. This is the way it was done in a couple other integrations, so I was attempting to be consistent, but moving forward, maybe something like the way I've updated the tests will be better. Updated in both email and phone spec.
- what is happening on line 169?
>> Leftover note to self. Removed.
- these tests cover successful data returned, but consider adding at least one for when looked-for data isn’t returned
>> in the failure tests, I test for what happens when no results come back. Do you mean something else? 

One last thing, about the naming: I wonder if it will be confusing that this company’s “email lookup” is meant to find an email (if I’m understanding this correctly), whereas other “email lookup” integrations are meant to find data by an email?
>> I see your point. Yes, you are correct, this company's email and phone lookup are both meant to find each based on lead data. Will need to think about this a bit more.

cgrayson commented:
>> I'm concerned that if I simplify this and there is a null value the field will be invalidated and I don't want the lookup to fail if that is the case.

Okay, but should be able to write a test to determine that for sure.

> perhaps all those if parsedBody?.datafinder?s could be factored out

I was thinking more of an `if parsedBody?.datafinder?` around those 4 statements.

> in describe ‘Validate’, could add a passes-everything case 
>> Not sure what you mean by this.

These tests ensure that missing data causes `validate()` to return an error message, but there's nothing to ensure that when all data is present, `validate()` returns nothing.

>> I used yes and no because that is the way the data comes back from Versium. Why do you think boolean is better?

Because booleans can be more easily worked with in flow rules, where the UI will automatically offer "is true" and "is false" as operators. A string would require the user to know that it returns "yes" or "no". Which your descriptions lists, that's good and not unworkable. But if we standardize, then users can learn that all "is xyz" attributes, from all integrations, work the same. Otherwise it could be "yes"/"no" here, "true"/"false" (as strings) somewhere else, etc. 

> what’s going on with the alternate_emails and alternate_phones arrays? those are populated with the given type, plus a timestamp in parens? 
>> yes, that is correct. Those are emails or phone numbers which are also associated with the lead data.

Aside from maybe straight delivery of that data to some CRM somewhere, that doesn't seem very useful to me. Leaving aside current limitations in array support, if they were actual `phone` and `email` types, then they'd have a better chance of being more actionable later in the flow (e.g., they would have sub-components, and could be tested for valid format). What the parenthetical timestamps mean or whether they're useful at all, I'm not sure.

>> maybe something like the way I've updated the tests will be better.

I think this is much easier to read, thanks. 

>> in the failure tests, I test for what happens when no results come back. Do you mean something else?

Not sure it applies, but will all the attributes in the fixture for the success tests always be returned? For example, if `EmailAddrUsable` isn't returned by Versium, what value does the integration put in `email_is_usable` (`null`, `undefined`, `''`)? If you do decide that's worth testing, then probably testing just one token attribute would be enough.

ros3bud commented:
Okay, but should be able to write a test to determine that for sure.
>> Test added in common_spec Validate description.

I was thinking more of an if parsedBody?.datafinder? around those 4 statements.
>> Ok, I see what you mean now. I can't wrap an if statement around all 4 of those statements because sometimes when datafinder is present, errors and or results may not be. For example, if there are no errors present, the errors key will not be there and if there is an error, results will not be there. I did add the conditional for dataFinder and numberOfResults though. I think that will work - thanks.

These tests ensure that missing data causes validate() to return an error message, but there's nothing to ensure that when all data is present, validate() returns nothing.
>> I thought that my tests for the request had that covered because when data is passed for any of those internally typed values, an error does not come up. But I see what you mean, in terms of making that more explicit and added tests to the validate function to make sure there are no errors when data is passed. 

Aside from maybe straight delivery of that data to some CRM somewhere, that doesn't seem very useful to me.
>> I see what you mean, but because the data is available, and given limited feedback about functionality needs, I made the decision to pass it to the customer, to be thorough. It's possible that the first result they get back will not be correct. This gives them the chance to look into alternates if they want to. I'll keep my ears open as this proceeds through QA and am happy to take it out if it turns out that it's officially unwanted data. Thanks for your feedback about that.  

will all the attributes in the fixture for the success tests always be returned? 
>>Yes, the attributes will be there, but no value will be returned if there are no alternates. Null test added for email and phone. 

I also got further feedback from Versium regarding EmailAddrUsable. It turns out that this will always be yes. This makes me think the response var should be taken out and I have done so. 

booleans can be more easily worked with in flow rules
>> makes sense, thanks. Updated.

cgrayson commented on src/common.coffee:
```diff
@@ -55,9 +55,11 @@ createRequest = (vars, service) ->
+    dataFinder = parsedBody.datafinder
+    numberOfResults = parsedBody.datafinder['num-results']
+  versiumError = parsedBody.datafinder.errors[0] if dataFinder?.errors?
+  results = parsedBody.datafinder.results[0] if dataFinder?.results?
```
I think you could still put these two lines inside the `if`, and then omit the `?`s at the end of `dataFinder`.

ros3bud replied:
Yes - I can omit the? - good point. However, unless you're seeing something I'm not, I don't think I can put those two lines inside the if. This is because sometimes datafinder is present, but results[] is not. Same with errors[].

cgrayson commented:
> I thought that my tests for the request had that covered because when data is passed for any of those internally typed values, an error does not come up.

In the tests, `request()` is being called explicitly; `validate()` is not involved there. It's only in the actual handler process that `validate()` is automatically called before `request()`.

cgrayson commented:
> I see what you mean, but because the data is available, and given limited feedback about functionality needs, I made the decision to pass it to the customer, to be thorough. It's possible that the first result they get back will not be correct. This gives them the chance to look into alternates if they want to.

Fair enough to get more input from the business side. My point is just that this _doesn't_ give anyone that much of a chance to use this. If someone is visually looking at appended data, or if the user builds the flow in a way that this `alternate_emails` field is passed into some plain-text field in the buyer's CRM, and that system's end-user knows to look there and what it means… then maybe. 

But there's no way to evaluate validity, or pass on to a verification service, or map onto `phone_2` or `phone_3`, etc.

cgrayson commented:
LGTM

ros3bud commented:
Thanks!
